rviz_default_plugins/src/rviz_default_plugins/tools/interaction/interaction_tool.cpp

```diff
@@ -0,0 +1,202 @@
+}
+
+void InteractionTool::activate()
+{
+  context_->getHandlerManager()->enableInteraction(true);
+  context_->getSelectionManager()->setTextureSize(1);
+}
+
+void InteractionTool::deactivate()
+{
+  context_->getHandlerManager()->enableInteraction(false);
+}
+
+void InteractionTool::updateFocus(const rviz_common::ViewportMouseEvent & event)
+{
+  const auto selection_manager = context_->getSelectionManager();
+  // Select exactly 1 pixel
+  selection_manager->select(
+    event.panel->getRenderWindow(),
+    event.x,
+    event.y,
+    event.x + 1,
+    event.y + 1,
+    rviz_common::interaction::SelectionManagerIface::SelectType::Add);
+
+  const rviz_common::interaction::M_Picked results = selection_manager->getSelection();
+
+  last_selection_frame_count_ = context_->getFrameCount();
+
+  rviz_common::InteractiveObjectPtr new_focused_object;
+
+  // look for a valid handle in the result.
+  auto result_it = results.begin();
+  if (result_it != results.end()) {
+    const rviz_common::interaction::Picked pick = result_it->second;
+    const auto handler = context_->getHandlerManager()->getHandler(pick.handle);
+    if (pick.pixel_count > 0 && handler) {
+      const rviz_common::InteractiveObjectPtr object = handler->getInteractiveObject().lock();
+      if (object && object->isInteractive()) {
+        new_focused_object = object;
+      }
+    }
+  }
+
+  // If the mouse has gone from one object to another, defocus the old
+  // and focus the new.
+  rviz_common::InteractiveObjectPtr new_obj = new_focused_object;
+  rviz_common::InteractiveObjectPtr old_obj = focused_object_.lock();
+  if (new_obj != old_obj) {
+    // Only copy the event contents here, once we know we need to use
+    // a modified version of it.
+    rviz_common::ViewportMouseEvent event_copy = event;
+    if (old_obj) {
+      event_copy.type = QEvent::FocusOut;
+      old_obj->handleMouseEvent(event_copy);
+    }
+
+    if (new_obj) {
+      event_copy.type = QEvent::FocusIn;
+      new_obj->handleMouseEvent(event_copy);
+    }
+  }
+
+  focused_object_ = new_focused_object;
+}
+
+int InteractionTool::processMouseEvent(rviz_common::ViewportMouseEvent & event)
+{
+  int flags = 0;
+
+  if (event.panel->contextMenuVisible()) {
+    return flags;
+  }
+
+  // make sure we let the vis. manager render at least one frame between selection updates
+  const bool need_selection_update = context_->getFrameCount() > last_selection_frame_count_;
+
+  // We are dragging if a button was down and is still down
+  Qt::MouseButtons buttons = event.buttons_down &
+    (Qt::LeftButton | Qt::RightButton | Qt::MidButton);
+  if (event.type == QEvent::MouseButtonPress) {
+    buttons &= ~event.acting_button;
+  }
+  const bool dragging = buttons != 0;
+
+  // unless we're dragging, check if there's a new object under the mouse
+  if (need_selection_update &&
+    !dragging &&
+    event.type != QEvent::MouseButtonRelease)
+  {
+    updateFocus(event);
+    flags = Render;
+  }
+
+  {
```

Comment:
If I see this correctly, this is a block to allow the weak pointer to be locked and unlocked correctly. Why not push this to its own function? This would break up the quite large method into smaller methods and enhance expressiveness.

Reply:
I've refactored the method by pulling out some of the logic into smaller methods: 5a5337c